CSS Nesting: nested selector matching

mdubet opened the pull request:
#### 8972793414d954d8dd57676c6f8f2298a001e73e
<pre>
CSS Nesting: nested selector matching
<a href="https://bugs.webkit.org/show_bug.cgi?id=249746">https://bugs.webkit.org/show_bug.cgi?id=249746</a>
rdar://103147183

Reviewed by Antti Koivisto.

This patch implements nested selector matching by
creating the flat equivalent of the nested selector
at rule set builder time.
All the following steps (matching, selector jit compilation,...)
should be automatically working because they will operate on regular flat selector,
without any parent selector inside them.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-003-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::simpleSelectorSpecificity):
(WebCore::CSSSelector::selectorText const):
(WebCore::CSSSelector::RareData::RareData):
(WebCore::CSSSelector::RareData::deepCopy const):
(WebCore::CSSSelector::CSSSelector):
(WebCore::CSSSelector::visitAllSimpleSelectors const):
(WebCore::CSSSelector::resolveNestingParentSelectors):
(WebCore::CSSSelector::hasExplicitNestingParent const):
* Source/WebCore/css/CSSSelector.h:
(WebCore::CSSSelector::selectorList):
(WebCore::CSSSelector::setNotLastInSelectorList):
(WebCore::CSSSelector::isFirstInTagHistory const):
(WebCore::CSSSelector::isLastInTagHistory const):
(WebCore::CSSSelector::setLastInTagHistory):
(WebCore::CSSSelector::tagHistory):
(WebCore::CSSSelector::CSSSelector): Deleted.
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::StyleRule):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::CSSParserSelector):
* Source/WebCore/css/parser/CSSParserSelector.h:
(WebCore::CSSParserSelector::selector):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeRelativeNestedSelector):
(WebCore::CSSSelectorParser::consumeNesting):
(WebCore::CSSSelectorParser::resolveNestingParent):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/style/RuleData.h:
(WebCore::Style::RuleData::selector const):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRule):
* Source/WebCore/style/RuleSetBuilder.h:

Canonical link: <a href="https://commits.webkit.org/258367@main">https://commits.webkit.org/258367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8543ac2d57fa72d08bdc968ca648bf0c18f4cf53

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111004 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171207 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105650 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1731 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94076 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108766 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107450 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8992 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92250 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/36580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90885 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23677 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78560 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4418 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25168 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4493 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1617 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44656 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5735 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6247 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+    }
+}
+
+static void visit(auto& selector, auto& apply)
```
Maybe something more informative like `visitAllSimpleSelectors`

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
     return count == b;
 }
 
+CSSSelector::CSSSelector(const CSSSelector& o)
```
o->other

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+
+void CSSSelector::flatten(const CSSSelectorList& parent)
+{
+    // ALWAYS_LOG_WITH_STREAM(stream << "should flatten selector " << selectorText() << " with parent " << parent.selectorsText());
```
some leftover logging code

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+
+    visit(*this, replaceParentSelector);
+    
+    // ALWAYS_LOG_WITH_STREAM(stream << "we are now flat: " << selectorText());
```
some leftover logging code

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+        visit(*next, apply); 
+}
+
+void CSSSelector::flatten(const CSSSelectorList& parent)
```
`flatten` is vague. I don't have a great suggestion but something along the lines `resolveNestingParentSelectors`

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+    // ALWAYS_LOG_WITH_STREAM(stream << "should flatten selector " << selectorText() << " with parent " << parent.selectorsText());
+
+    auto replaceParentSelector = [&] (CSSSelector& selector) {
+        if (selector.match() == CSSSelector::PseudoClass && selector.pseudoClassType() == CSSSelector::PseudoClassParent) {
```
Not from this patch but `PseudoClassParent` could use a better name. Something like PseudoClassNestingParent so it is clear what sort of "parent" we are talking about. The spec calls & a "nesting selector".

mdubet replied:
Let's use "NestingParent" everywhere 👍

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+    // ALWAYS_LOG_WITH_STREAM(stream << "we are now flat: " << selectorText());
+}
+
+bool CSSSelector::hasExplicitParent() const
```
Also naming

anttijk commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -250,6 +253,7 @@ struct PossiblyQuotedIdentifier {
         // Selectors are kept in an array by CSSSelectorList.
         // The next component of the selector is the next item in the array.
         const CSSSelector* tagHistory() const { return m_isLastInTagHistory ? nullptr : this + 1; }
+        CSSSelector* tagHistory() { return m_isLastInTagHistory ? nullptr : this + 1; }
```
Can this be private? Generally no one should be mutating selectors outside some very specific context.

anttijk commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -70,7 +70,6 @@ CSSSelectorList::CSSSelectorList(Vector<std::unique_ptr<CSSParserSelector>>&& se
             if (current != first)
                 m_selectorArray[arrayIndex].setNotFirstInTagHistory();
             current = current->tagHistory();
-            ASSERT(!m_selectorArray[arrayIndex].isLastInSelectorList());
```
Why are you removing this assert?

mdubet replied:
This assert is a bit weird (it works only because the default ctor write `false` in this field, and it assumes this field has never been modified). It should actually be a `setNotLastInSelectorList()` (which is overwritten for the last selector in the last line of the function) because we don't care about what is there already, we just need it to be false for all but the last one after construction.

I put it back, except for the last element.

anttijk commented on Source/WebCore/css/CSSSelectorList.h:
```diff
@@ -47,6 +47,7 @@ class CSSSelectorList {
     const CSSSelector* first() const { return m_selectorArray.get(); }
     static const CSSSelector* next(const CSSSelector*);
     const CSSSelector* selectorAt(size_t index) const { return &m_selectorArray[index]; }
+    CSSSelector* selectorAt(size_t index) { return &m_selectorArray[index]; }
```
I don't see this accessor being used.

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -945,4 +964,82 @@ bool CSSSelector::RareData::matchNth(int count)
+    auto replaceParentSelector = [&] (CSSSelector& selector) {
+        if (selector.match() == CSSSelector::PseudoClass && selector.pseudoClassType() == CSSSelector::PseudoClassParent) {
+            selector.setMatch(Match::PseudoClass);
+            selector.setPseudoClassType(PseudoClassType::PseudoClassIs);
+            selector.setSelectorList(makeUnique<CSSSelectorList>(parent));
```
This is reasonable for now but there are common case where desugaring can avoid inserting an `:is` and just insert the selector itself for better performance. Might want to add comment about it.

anttijk commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -250,6 +252,43 @@ Ref<StyleRule> StyleRule::copy() const
     return adoptRef(*new StyleRule(*this));
 }
 
+void StyleRule::initializeFlatSelectorList(const CSSSelectorList& parentFlatSelectorList)
```
I don't like the use of word "flat" here. It doesn't appear to be a spec term either. `resolveAndCacheNestedParentSelector` or something?

StyleRule is not necessarily the right place for the function that does the desugaring. Somewhere around CSSSelector/SelectorList/Parser would be more natural. It would be better to just move in a fully resolved selector.

anttijk commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -250,6 +252,43 @@ Ref<StyleRule> StyleRule::copy() const
+    auto selector = copiedSelectorList.first();
+    while (selector) {
+        if (selector->hasExplicitParent()) {
+            const_cast<CSSSelector*>(selector)->flatten(parentFlatSelectorList);
```
Probably better to expose mutable accessors (at least using friend) if we really want to mutate CSSSelectoLlists. Generally it would be better to mutate things as CSSParserSelector vector and then construct a CSSSelectorList at the end.

mdubet replied:
I've added a FIXME to refactor it (build a CSSParserSelector, resolve it, and build a CSSSelectorList from it) in a following patch.

anttijk commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -147,6 +154,7 @@ class StyleRule final : public StyleRuleBase {
     CSSSelectorList m_selectorList;
 
     // CSS Nesting
+    std::optional<CSSSelectorList> m_flatSelectorList;
```
While optional is semantically nice it does use more space (a single bit, but alignment makes it bigger) . Can empty CSSSelectorList represent the null case?

This could be marked 'mutable' so the setter can be const. This is common for cache-like fields.

A better name wouldn't need  the comment.

mdubet replied:
It's possible, but we need to remove an assert which prevent copying an empty CSSSelectorList.

anttijk commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -110,6 +110,12 @@ class StyleRule final : public StyleRuleBase {
     ~StyleRule();
 
     const CSSSelectorList& selectorList() const { return m_selectorList; }
+    const CSSSelectorList& processedSelectorList() const
```
`resolvedSelectorList`?
`selectorListWithResolvedNesting`?

mdubet replied:
`resolvedSelectorList` ✅

anttijk commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -203,20 +203,29 @@ void RuleSetBuilder::addRulesFromSheetContents(const StyleSheetContents& sheet)
+    if (parentFlatSelectorList) {
+        // This is some kind of cache, let's consider this "inoffensive mutability".
+        StyleRule& mutableRule = const_cast<StyleRule&>(rule);
+        mutableRule.initializeFlatSelectorList(*parentFlatSelectorList);
```
See the comments about using 'mutable' above. When things are cache-like it can be helpful to just call them "cache".

anttijk commented on Source/WebCore/style/RuleSetBuilder.h:
```diff
@@ -34,7 +34,7 @@ class RuleSetBuilder {
 
     void addRulesFromSheet(const StyleSheetContents&, const MQ::MediaQueryList& sheetQuery = { });
-    void addStyleRule(const StyleRule&);
+    void addStyleRule(const StyleRule&, const CSSSelectorList* = nullptr);
```
Would be good to have a name here since it is non-obvious what the selector list is for.

anttijk reviewed (approved):
Please watch perf-bots when landing. I don't think it matter that StyleRule gets bigger but maybe it does?

mdubet committed:
CSS Nesting: nested selector matching
https://bugs.webkit.org/show_bug.cgi?id=249746
rdar://103147183

Reviewed by Antti Koivisto.

This patch implements nested selector matching by
creating the flat equivalent of the nested selector
at rule set builder time.
All the following steps (matching, selector jit compilation,...)
should be automatically working because they will operate on regular flat selector,
without any parent selector inside them.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-003-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::simpleSelectorSpecificity):
(WebCore::CSSSelector::selectorText const):
(WebCore::CSSSelector::RareData::RareData):
(WebCore::CSSSelector::RareData::deepCopy const):
(WebCore::CSSSelector::CSSSelector):
(WebCore::CSSSelector::visitAllSimpleSelectors const):
(WebCore::CSSSelector::resolveNestingParentSelectors):
(WebCore::CSSSelector::hasExplicitNestingParent const):
* Source/WebCore/css/CSSSelector.h:
(WebCore::CSSSelector::selectorList):
(WebCore::CSSSelector::setNotLastInSelectorList):
(WebCore::CSSSelector::isFirstInTagHistory const):
(WebCore::CSSSelector::isLastInTagHistory const):
(WebCore::CSSSelector::setLastInTagHistory):
(WebCore::CSSSelector::tagHistory):
(WebCore::CSSSelector::CSSSelector): Deleted.
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::StyleRule):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::CSSParserSelector):
* Source/WebCore/css/parser/CSSParserSelector.h:
(WebCore::CSSParserSelector::selector):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeRelativeNestedSelector):
(WebCore::CSSSelectorParser::consumeNesting):
(WebCore::CSSSelectorParser::resolveNestingParent):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/style/RuleData.h:
(WebCore::Style::RuleData::selector const):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRule):
* Source/WebCore/style/RuleSetBuilder.h:

Canonical link: https://commits.webkit.org/258367@main
